### PR TITLE
feat(js/plugins): added custom dimensions parameter to text-embedding

### PR DESCRIPTION
**Pull request by river0g**
How to use
```typescript
ai.embed({
  embedder: textEmbedding005,
  content: 'String to be embedded',
  options: {
    outputDimensionality: 500 // Between 1 and 768, inclusive
  }
})
```

Checklist (if applicable):
- [x] PR title is following https://www.conventionalcommits.org/en/v1.0.0/
- [x] Tested (manually, unit tested, etc.)

### CODE REVIEW

File: js/plugins/googleai/src/embedder.ts [hunk 1]
```diff
@@ -36,6 +36,14 @@ export const GeminiEmbeddingConfigSchema = z.object({
...
   title: z.string().optional(),
   version: z.string().optional(),
+  /**
+   * The `dimensions` parameter allows you to specify the dimensionality of the embedding output.
+   * By default, the model generates embeddings with 768 dimensions. Models such as
+   * `text-embedding-004`, `text-embedding-005`, and `text-multilingual-embedding-002`
+   * allow the output dimensionality to be adjusted between 1 and 768.
+   * By selecting a smaller output dimensionality, users can save memory and storage space, leading to more efficient computations.
+   **/
+  dimensions: z.number().min(1).max(768).optional(),
```
Comment (pavelgj) [inline js/plugins/googleai/src/embedder.ts:46]: the underlying option is called `outputDimensionality`... I like `dimensions` better, but I still think we should stick as close to to API option naming as possible. It's easier to cross-reference these options with API documentation...

File: js/plugins/googleai/src/embedder.ts [hunk 2]
```diff
@@ -69,6 +77,8 @@ export const SUPPORTED_MODELS = {
   'text-embedding-004': textEmbedding004,
 };
 
+const customDimensionsSupportedModelNames = [textEmbedding004.name];
```
Comment (pavelgj) [inline js/plugins/googleai/src/embedder.ts:80]: I don't think this is necessary. I don't think we want to maintain this mapping. Let the API return errors if the option is not supported.

File: js/plugins/vertexai/src/embedder.ts [hunk 3]
```diff
@@ -39,6 +39,14 @@ export const VertexEmbeddingConfigSchema = z.object({
...
   location: z.string().optional(),
   version: z.string().optional(),
+  /**
+   * The `dimensions` parameter allows you to specify the dimensionality of the embedding output.
+   * By default, the model generates embeddings with 768 dimensions. Models such as
+   * `text-embedding-004`, `text-embedding-005`, and `text-multilingual-embedding-002`
+   * allow the output dimensionality to be adjusted between 1 and 768.
+   * By selecting a smaller output dimensionality, users can save memory and storage space, leading to more efficient computations.
+   **/
+  dimensions: z.number().min(1).max(768).optional(),
```
Comment (pavelgj) [inline js/plugins/vertexai/src/embedder.ts:49]: same thing on the naming -- let's call it `outputDimensionality` for consistency

File: js/plugins/vertexai/src/embedder.ts [hunk 4]
```diff
@@ -83,6 +91,12 @@ export const SUPPORTED_EMBEDDER_MODELS: Record<string, EmbedderReference> = {
   // ]),
 };
 
+const customDimensionsSupportedModelNames = [
```
Comment (pavelgj) [inline js/plugins/vertexai/src/embedder.ts:94]: same thing here, let's drop this mapping.

File: js/plugins/googleai/src/embedder.ts (hunk 2 above)
Comment (river0g) [reply js/plugins/googleai/src/embedder.ts:80]: Indeed, based on your comment, I also feel that an API error would be the more natural approach.
I thought it was important to ensure that the issue is not hidden when the option is applied to an embedding model that does not support it.
Thank you for your comment.

File: js/plugins/vertexai/src/embedder.ts [hunk 5]
```diff
@@ -145,7 +153,8 @@ export function defineVertexAIEmbedder(
             task_type: options?.taskType,
             title: options?.title,
           };
-        })
+        }),
+        { outputDimensionality: options?.outputDimensionality }
```
Comment (pavelgj) [inline js/plugins/vertexai/src/embedder.ts:157]: why is not along with other options (lines 152-154)?

Comment (river0g) [reply js/plugins/vertexai/src/embedder.ts:157]: Thank you for your question.
Based on the API interface, it seems that `outputDimensionality` cannot be included under `instances`, and must be placed under `parameters`.
(documentation here: [Get text embeddings](https://cloud.google.com/vertex-ai/generative-ai/docs/embeddings/get-text-embeddings#get_text_embeddings_for_a_snippet_of_text))

After verifying the actual behavior, I found that when `outputDimensionality` is placed under `instances` (lines 152-154), it is ignored, and the default dimensionality is returned.

Comment (pavelgj) [reply js/plugins/vertexai/src/embedder.ts:157]: Perfect! Just making sure.

**Review by pavelgj — APPROVED** [review]

**Pull request merged**